src/scanner.rs

```diff
@@ -0,0 +1,262 @@
+use crate::error::error;
+use crate::token::{Token, TokenType};
+use crate::util::*;
+
+use std::iter::Peekable;
+use unicode_segmentation::{Graphemes, UnicodeSegmentation};
+
+pub struct Scanner<'a> {
+    pub source: Peekable<Graphemes<'a>>,
+    pub tokens: Vec<Token>,
+
+    pub line: usize,
+}
+
+impl<'a> Scanner<'a> {
+    pub fn new(s: &'a str) -> Self {
+        Self {
+            source: UnicodeSegmentation::graphemes(s, true).peekable(),
+            line: 0,
+            tokens: Vec::new(),
+        }
+    }
+
+    fn error(&self, message: &str) {
+        error(self.line, message);
+    }
+
+    fn add_token(&mut self, token_type: TokenType) {
+        self.tokens.push(Token {
+            token_type,
+            line: self.line,
+        })
+    }
+
+    fn peek(&mut self) -> &str {
+        let c = self.source.peek();
+
+        if c.is_none() {
+            EOF
+        } else {
+            *c.unwrap()
+        }
+    }
+
+    fn consume(&mut self) {
+        let c = self.source.next();
+
+        if c.is_none() {
+            error(self.line, "unexpected end of file");
+        }
+    }
+
+    fn consume_and_return(&mut self) -> &str {
+        let c = self.source.next();
+
+        if c.is_none() {
+            error(self.line, "unexpected end of file");
+        }
+
+        c.unwrap()
+    }
+
+    fn consume_if_next(&mut self, condition: fn(&str, &str) -> bool) -> bool {
+        if self.peek() == EOF {
+            false
```

Comment:
```suggestion
            return false
```
get rid of the extra indent